File: decorator.cs
using System;

class Coffee {
    public virtual int Cost() => 5;
}

class CoffeeDecorator : Coffee {
    protected Coffee coffee;
    public CoffeeDecorator(Coffee coffee) {
        this.coffee = coffee;
    }

    public override int Cost() => coffee.Cost();
}

class MilkDecorator : CoffeeDecorator {
    public MilkDecorator(Coffee coffee) : base(coffee) { }

    public override int Cost() => base.Cost() + 2;
}

class SugarDecorator : CoffeeDecorator {
    public SugarDecorator(Coffee coffee) : base(coffee) { }

    public override int Cost() => base.Cost() + 1;
}

class Program {
    static void Main() {
        Coffee coffee = new Coffee();
        Console.WriteLine(coffee.Cost());  // 5

        Coffee milk = new MilkDecorator(coffee);
        Console.WriteLine(milk.Cost());    // 7

        Coffee sugar = new SugarDecorator(coffee);
        Console.WriteLine(sugar.Cost());   // 6

        Coffee milkAndSugar = new SugarDecorator(milk);
        Console.WriteLine(milkAndSugar.Cost());  // 8
    }
}

File: decorator.kt
open class Coffee {
    open fun cost(): Int = 5
}

open class CoffeeDecorator(private val coffee: Coffee) : Coffee() {
    override fun cost(): Int = coffee.cost()
}

class MilkDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int = super.cost() + 2
}

class SugarDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int = super.cost() + 1
}

fun main() {
    val coffee = Coffee()
    println(coffee.cost())  // 5

    val milk = MilkDecorator(coffee)
    println(milk.cost())    // 7

    val sugar = SugarDecorator(coffee)
    println(sugar.cost())   // 6

    val milkAndSugar = SugarDecorator(milk)
    println(milkAndSugar.cost())  // 8
}
